docs/Toy/toy/parser.py

```diff
@@ -0,0 +1,437 @@
+
+            dims += first
+
+        return LiteralExprAST(openBracket.loc, values, dims)
+
+    def parseParenExpr(self) -> ExprAST:
+        "parenexpr ::= '(' expression ')'"
+        self.pop_pattern('(')
+        v = self.parseExpression()
+        self.pop_pattern(')')
+        return v
+
+    def parseIdentifierExpr(self):
+        '''
+        identifierexpr
+        ::= identifier
+        ::= identifier '(' expression ')'
+        '''
+        name = self.pop_token(IdentifierToken)
+        if not self.check('('):
+            # Simple variable ref.
+            return VariableExprAST(name.loc, name.text)
+
+        # This is a function call.
+        self.pop_pattern('(')
+        args: List[ExprAST] = []
+        while True:
+            args.append(self.parseExpression())
+            if self.check(')'):
+                break
+            self.pop_pattern(',')
+        self.pop_pattern(')')
+
+        if name.text == 'print':
+            # It can be a builtin call to print
+            if len(args) != 1:
+                self.parseError('<single arg>', 'as argument to print()')
+
+            return PrintExprAST(name.loc, args[0])
+
+        return CallExprAST(name.loc, name.text, args)
+
+    def parsePrimary(self) -> ExprAST | None:
+        '''
+        primary
+        ::= identifierexpr
+        ::= numberexpr
+        ::= parenexpr
+        ::= tensorliteral
+        '''
+        current = self.tokens[self.pos]
+        if isinstance(current, IdentifierToken):
+            return self.parseIdentifierExpr()
+        elif isinstance(current, NumberToken):
+            return self.parseNumberExpr()
+        elif current.text == '(':
+            return self.parseParenExpr()
+        elif current.text == '[':
+            return self.parseTensorLiteralExpr()
+        elif current.text == ';':
+            return None
+        elif current.text == '}':
+            return None
+        else:
+            self.parseError('expression or one of `;`, `}`')
+
+    def parsePrimaryNotNone(self) -> ExprAST:
+        '''
+        primary
+        ::= identifierexpr
+        ::= numberexpr
+        ::= parenexpr
+        ::= tensorliteral
+        '''
+        current = self.tokens[self.pos]
+        if isinstance(current, IdentifierToken):
+            return self.parseIdentifierExpr()
+        elif isinstance(current, NumberToken):
+            return self.parseNumberExpr()
+        elif current.text == '(':
+            return self.parseParenExpr()
+        elif current.text == '[':
+            return self.parseTensorLiteralExpr()
+        else:
+            self.parseError('expression')
+
+    def parseBinOpRHS(self, exprPrec: int, lhs: ExprAST) -> ExprAST:
+        '''
+        Recursively parse the right hand side of a binary expression, the ExprPrec
+        argument indicates the precedence of the current binary operator.
+
+        binoprhs ::= ('+' primary)*
+        '''
+        # If this is a binop, find its precedence.
```

Comment:
BinOp

Reply:
I read it as a reference to the rule in the grammar, spelled `binop`, as opposed to a reference to the class, spelled `BinOp`, so I don't think it's incorrect